website/addons/s3/static/s3NodeConfig.js

```diff
@@ -0,0 +1,349 @@
+'use strict';
+var ko = require('knockout');
+require('knockout-punches');
+require('knockout-mapping');
+var $ = require('jquery');
+var bootbox = require('bootbox');
+var Raven = require('raven-js');
+
+var $osf = require('js/osfHelpers');
+
+ko.punches.enableAll();
+
+var noop = function() {};
+
+var ViewModel = function(url, selector) {
+    var self = this;
+
+    self.url = url;
+    self.selector = selector;
+
+    self.nodeHasAuth = ko.observable(false);
+    self.userHasAuth = ko.observable(false);
+    self.userIsOwner = ko.observable(false);
+    self.ownerName = ko.observable('');
+
+    self.urls = ko.observable({});
+    self.loadedSettings = ko.observable(false);
+    self.bucketList = ko.observableArray([]);
+    self.loadedBucketList = ko.observable(false);
+    self.currentBucket = ko.observable('');
+    self.selectedBucket = ko.observable('');
+
+    self.accessKey = ko.observable('');
+    self.secretKey = ko.observable('');
+
+    self.loading = ko.observable(false);
+    self.creating = ko.observable(false);
+    self.creatingCredentials = ko.observable(false);
+
+    self.message = ko.observable('');
+    self.messageClass = ko.observable('text-info');
+
+    self.showSelect = ko.observable(false);
+
+    self.showSettings = ko.pureComputed(function() {
+        return self.nodeHasAuth();
+    });
+    self.disableSettings = ko.pureComputed(function() {
+        return !(self.userHasAuth() && self.userIsOwner());
+    });
+    self.showNewBucket = ko.pureComputed(function() {
+        return self.userHasAuth() && self.userIsOwner();
+    });
+    self.showImport = ko.pureComputed(function() {
+        return self.userHasAuth() && !self.nodeHasAuth();
+    });
+    self.showCreateCredentials = ko.pureComputed(function() {
+        return !self.nodeHasAuth() && !self.userHasAuth();
+    });
+    self.canChange = ko.pureComputed(function() {
+        return self.userIsOwner() && self.nodeHasAuth();
+    });
+    self.allowSelectBucket = ko.pureComputed(function() {
+        return (self.bucketList().length > 0 || self.loadedBucketList()) && (!self.loading());
+    });
+};
+
+ViewModel.prototype.toggleSelect = function() {
+    this.showSelect(!this.showSelect());
+    if (!this.loadedBucketList()) {
+        return this.fetchBucketList();
+    }
+    return new $.Deferred().promise();
```

Comment:
`toggleSelect` and `fetchBucketList` could use refactoring. `fetchBucketList` does more than fetching; it also sets attributes/observables. The "empty" promise here also isn't very useful.

What if `fetchBucketList` only fetched buckets (and handled unexpected errors), and toggleSelect just set the `bucketList`?

Something like:

```js

ViewModel.prototype.fetchBucketList = function() {
    var ret = $.Deferred();
    var self = this;
    if (self.loadedBucketList()) {
        // resolve to cached bucketList
        ret.resolve(self.bucketList());
    } else{
        $.ajax({
            url: self.urls().bucket_list,
            type: 'GET',
            dataType: 'json',
        }).done(function(response) {
            self.loadedBucketList(true);
            ret.resolve(response.buckets);
        }).fail(function(xhr, status, error) {
            var message = 'Could not retrieve list of S3 buckets at' +
                'this time. Please refresh the page. If the problem persists, email ' +
                '<a href="mailto:support@osf.io">support@osf.io</a>.';
            self.changeMessage(message, 'text-warning');
            Raven.captureMessage('Could not GET s3 bucket list', {
                url: self.urls().bucketList,
                textStatus: status,
                error: error
            });
            ret.reject(xhr, status, error);
        });
    }
    return ret.promise();
};

ViewModel.prototype.togglePicker = function() {
    return this.fetchBucketList()
        .done(function(buckets) {
            self.bucketList(buckets);
            self.selectedBucket(self.currentBucket());
        });
}
```